aten/src/ATen/native/LinearAlgebra.cpp

```diff
@@ -515,6 +515,368 @@ Tensor& matmul_out(Tensor &result, const Tensor & tensor1, const Tensor & tensor
+  compute_Ts = {
+    compute_T1, compute_T2, compute_T4,
+    compute_T8<scalar_t>, compute_T12<scalar_t>
+  };
+
+  for (int i = 0; i < total_n_degs - 1; ++i) {
+    if ((norm <= thetas[i]).all().template item<bool>()) {
+      return compute_Ts[i](a);
+    }
+  }
+
+  // Scale
+  auto s = at::max(at::zeros_like(norm),
+              at::ceil(at::log2(norm / thetas[total_n_degs - 1]))).to(at::kLong);
+  auto pow2s = at::pow(2, s);
+  auto a_scaled = a / pow2s.unsqueeze(-1).unsqueeze(-1);
+
+  // Square
+  return matrix_power(compute_T18<scalar_t>(a_scaled), pow2s);
+}
+
+// matrix exponential
+Tensor mexp(const Tensor& a) {
+  if (a.scalar_type() == at::ScalarType::Float) {
+    constexpr std::array<float, total_n_degs> thetas_float = {
+      1.192092800768788e-07, // deg 1
+      5.978858893805233e-04, // deg 2
+      5.116619363445086e-02, // deg 4
+      5.800524627688768e-01, // deg 8
+      1.461661507209034e+00, // deg 12
+      3.010066362817634e+00  // deg 18
+    };
+    return mexp_impl<float>(a, thetas_float);
+  }
+  else { // if Double
+    constexpr std::array<double, total_n_degs> thetas_double = {
+      2.220446049250313e-16, // deg 1
+      2.580956802971767e-08, // deg 2
+      3.397168839976962e-04, // deg 4
+      4.991228871115323e-02, // deg 8
+      2.996158913811580e-01, // deg 12
+      1.090863719290036e+00  // deg 18
+    };
+    return mexp_impl<double>(a, thetas_double);
+  }
+}
+
+// Based on:
+//
+// Mathias, Roy. 
+// “A Chain Rule for Matrix Functions and Applications.”
+// SIAM J. Matrix Anal. Appl. 17 (1996): 610-620.
+//
+template <typename func_t>
+Tensor backward_analytic_function_of_a_matrix(
+    const Tensor& self, const Tensor& grad,
+    const func_t& function_of_a_matrix
+  ) {
+  auto self_transposed = self.transpose(-2, -1);
+  auto self_transposed_sizes = self_transposed.sizes().vec();
+  self_transposed_sizes[self.dim() - 2] <<= 1;
+  self_transposed_sizes[self.dim() - 1] <<= 1;
+
+  auto n = self_transposed.size(-1);
+  auto meta_grad = at::zeros(self_transposed_sizes, grad.options());
+  meta_grad.narrow(-2, 0, n).narrow(-1, 0, n).copy_(self_transposed);
+  meta_grad.narrow(-2, n, n).narrow(-1, n, n).copy_(self_transposed);
+  meta_grad.narrow(-2, 0, n).narrow(-1, n, n).copy_(grad);
+
+  auto grad_input = function_of_a_matrix(meta_grad)
+    .narrow(-2, 0, n).narrow(-1, n, n);
+  return grad_input;
+}
+
+};
+
+// Computes the matrix exponential for a given batch of squared matrices.
+// The implementaion is based on:
+//
+// Bader, P.; Blanes, S.; Casas, F.
+// Computing the Matrix Exponential with an Optimized Taylor Polynomial Approximation.
+// Mathematics 2019, 7, 1174.
+//
+Tensor matrix_exp(const Tensor& a) {
+  TORCH_CHECK(a.dim() >= 2 
+          && (at::isFloatingType(a.scalar_type()) 
+           || at::isComplexType(a.scalar_type())),
+              "matrix_exp(", a.scalar_type(), "{", a.sizes(), "}): expected a tensor "
+              "of floating types with dim at least 2");
```

Comment:
nit - `...of floating or complex dtype with ...`

Reply:
Hmm, none of the algorithms dispatching over floating types with complex mention complex in the checks. Does PyTorch support or plans to support complex numbers not over floating point types?

Reply:
yeah the error messages for other functions should be updated at some point. torch doesn't classify complex tensors as floating point, i.e. `complex_tensor.is_floating_point()` returns `false`